backend/src/services/BaseService.ts

```diff
@@ -0,0 +1,117 @@
+
+  const bundle = makeTransactionResponseBundle(requestResults, res, baseVersion);
+  logger.info('Transaction bundle successfully uploaded.');
+  return bundle;
+}
+
+/**
+ * Uploads Library and Measure resources from a transaction bundle to the server.
+ */
+async function uploadResourcesFromBundle(entries: DetailedEntry[]) {
+  logger.info(`Inserting Measure and Library resources from transaction bundle`);
+  const scrubbedEntries = replaceReferences(entries);
+
+  const requestsArray = scrubbedEntries.map(async entry => {
+    const { method } = entry.request as fhir4.BundleEntryRequest;
+    return insertBundleResources(entry, method);
+  });
+  return Promise.all(requestsArray);
+}
+
+/**
+ * Inserts bundle Library or Measure resources into the database through create or update.
+ */
+async function insertBundleResources(entry: DetailedEntry, method: string) {
+  if (entry.resource?.resourceType === 'Library' || entry.resource?.resourceType === 'Measure') {
+    if (method === 'POST') {
+      entry.resource.id = uuidv4();
+      const { id } = await createResource(entry.resource, entry.resource.resourceType);
+      if (id !== null) {
+        entry.status = 201;
+        entry.statusText = 'Created';
+      }
+    }
+    if (method === 'PUT') {
+      if (entry.resource.id) {
+        const { id, created } = await updateResource(entry.resource.id, entry.resource, entry.resource.resourceType);
+        if (created === true) {
+          entry.status = 201;
+          entry.statusText = 'Created';
+        } else if (id != null && created === false) {
+          entry.status = 200;
+          entry.statusText = 'OK';
+        }
+      } else {
+        throw new BadRequestError(`Requests of type PUT must have an id associated with the resource.`);
+      }
+    } else {
+      throw new BadRequestError(
+        `Expected requests of type PUT or POST, received ${method} for ${entry.resource?.resourceType}/${entry.resource?.id}`
+      );
+    }
+  } else {
+    throw new BadRequestError('Resource in the entry must be of type Measure or Library.');
+  }
+  return entry;
+}
+
+/**
+ * Creates transaction-response bundle.
+ */
+function makeTransactionResponseBundle(
+  requestResults: DetailedEntry[],
+  res: any,
+  baseVersion: keyof typeof constants.VERSIONS
+) {
+  logger.info('Compiling transaction response bundle');
+  const Bundle = resolveSchema(baseVersion, 'bundle');
+  const bundle = new Bundle({ type: 'transaction-response', id: uuidv4() });
+  bundle.link = {
+    url: `${res.req.protocol}://${path.join(res.req.get('host'), res.req.baseUrl)}`,
+    relation: 'self'
+  };
+
+  const entries: fhir4.BundleEntry[] = [];
+  requestResults.forEach(result => {
+    const entry = new Bundle({ response: { status: `${result.status} ${result.statusText}` } });
+    if (result.status === 200 || result.status === 201) {
+      entry.response.location = `${baseVersion}/${result.resource?.resourceType}/${result.resource?.id}`;
+    }
```

Comment:
Hmm I am a little confused about this and would love to discuss more. I was under the impression that for transaction bundles, if anything goes wrong with the requests (aka not 200 or 201), then an error would be thrown. No transaction-response bundle will be returned if any of the requests fail, right? At the end of that part of the spec, it says "This outcome is not used for error responses in batch/transaction, only for hints and warnings. In a batch operation, the error will be in Bundle.entry.response, and for transaction, there will be a single OperationOutcome instead of a bundle in the case of an error." Idk I am tired maybe I am totally misunderstanding this.

Reply:
I think this is a niche of transactions that doesn't super matter here or in the test server tbh. We can just not include outcome. I don't think we need to worry about bundles successfully uploading/updating only _parts_ of the transaction. We can just fail if any resources fail. That seems to be what HAPI does:

![image](https://user-images.githubusercontent.com/16297930/230468845-c91f4412-d11a-4ff8-b4b7-4cfdab47b505.png)

I don't remember doing that behavior in deqm-test-server or why we did it, but maybe we got it wrong. Doesn't really matter at this point